SSA/Experimental/IntrinsicAsymptotics.lean

```diff
@@ -93,47 +99,344 @@ mk_ex 120
+  let newProgram := addLetsToProgram newLets newProgram
+
+  some newProgram
+
+def rewriteAt' (inputProg : Com) (depth : Nat) (lets: Lets) (rewrite: ExprRec × ExprRec) : Option Com :=
+  match inputProg with
+    | .ret _ => none
+    | .let _ expr body =>
+        let lets := lets.push expr
+        if depth = 0 then
+           applyRewrite lets body rewrite
+        else
+           rewriteAt' body (depth - 1) lets rewrite
+
+def rewriteAt (inputProg : Com) (depth : Nat) (rewrite: ExprRec × ExprRec) : Option Com :=
+    rewriteAt' inputProg depth #[] rewrite 
+
+def rewrite (inputProg : Com) (depth : Nat) (rewrite: ExprRec × ExprRec) : Com :=
+    let x := rewriteAt inputProg depth rewrite 
+    match x with
+      | none => inputProg
+      | some y => y
 
 def Expr.denote : Expr → State → Value
-| .nat n, _ => .nat n
+| .cst n, _ => .nat n
 | .add a b, l => let a_val := get_nat (l.get! a)
                  let b_val := get_nat (l.get! b)
                  Value.nat (a_val + b_val)
 
 def Com.denote : Com → State → Value
-| .ret _ e, l => e.denote l
+| .ret v, l => l.get! v
 | .let _ e body, l => denote body ((e.denote l) :: l)
 
+def denote (p: Com) : Value :=
+  p.denote []
+
+def ExprRec.denote : ExprRec → State → Value
+| .cst n, _ => .nat n
+| .add a b, s => let a_val := get_nat (a.denote s)
+                 let b_val := get_nat (b.denote s)
+                 Value.nat (a_val + b_val)
+| .var v, s => s.get! v
+
+theorem shifting:
+denote (addLetsToProgram lets (shiftBy p n)) = denote p := sorry
+
+
+theorem letsTheorem 
+ (matchExpr : ExprRec) (lets : Lets)
+ (h1: matchVar lets (lets.size-1) matchExpr m₀ = some m)
+ (hlets: lets.size > 0)
+ (hm₀: denote (addLetsToProgram lets (Com.ret 0)) =
+       denote (addLetsToProgram (applyMapping matchExpr m₀ lets).1 (Com.ret 0))):
+
+   denote (addLetsToProgram (lets) (Com.ret 0)) =
+   denote (addLetsToProgram ((applyMapping matchExpr m lets).1) (Com.ret 0)) := by
+      induction matchExpr
+      unfold applyMapping
+      simp
+      case cst =>
+        simp [matchVar] at h1
+        split at h1
+        rename_i x n heq
+        · rw [hm₀]
+          split at h1 <;> unfold applyMapping <;> simp
+        · contradiction
+      
+      case add a b a_ih b_ih =>
+         
+        rw [hm₀]
+        induction lets.size
+        simp_all
+
+        unfold addLetsToProgram
+        unfold addLetsToProgram'
+        
+
+        split at h1
+        rename_i x a' b' heq
+
+        · simp_all
+          
+
+          sorry         
+        . contradiction
+
+      case var =>
+        simp [matchVar] at h1
+        split at h1
+        rename_i x n heq
+        · split at h1 <;> rw [hm₀] <;> unfold applyMapping <;> simp
+        · erw [hm₀]
+
+theorem rewriteAtCorrect 
+  (p : Com) (pos: Nat) (rwExpr : ExprRec × ExprRec) 
+  (rewriteCorrect : ∀ s : State, rwExpr.1.denote s = rwExpr.2.denote s)
+  (lets : Lets) (successful : rewriteAt' p pos lets rwExpr = some p'):
```

Comment:
I wonder if it might be beneficial to have `pos` be `Fin (p.size)`, where `p.size` is the number of let-bindings. That is a relatively mild addition of dependent typing, but should let us get rid of this `successful` hypothesis, potentially simplifying proofs